internal/controllers/machinedeployment/machinedeployment_status_test.go

```diff
@@ -0,0 +1,1185 @@
+			expectCondition: metav1.Condition{
+				Type:    clusterv1.MachineDeploymentAvailableV1Beta2Condition,
+				Status:  metav1.ConditionFalse,
+				Reason:  clusterv1.MachineDeploymentNotAvailableV1Beta2Reason,
+				Message: "3 available replicas, at least 4 required (spec.strategy.rollout.maxUnavailable is 1)",
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			setAvailableCondition(ctx, tt.machineDeployment, tt.getAndAdoptMachineSetsForDeploymentSucceeded)
+
+			condition := v1beta2conditions.Get(tt.machineDeployment, clusterv1.MachineDeploymentAvailableV1Beta2Condition)
+			g.Expect(condition).ToNot(BeNil())
+			g.Expect(*condition).To(v1beta2conditions.MatchCondition(tt.expectCondition, v1beta2conditions.IgnoreLastTransitionTime(true)))
+		})
+	}
+}
+
+func Test_setScalingUpCondition(t *testing.T) {
+	defaultMachineDeployment := &clusterv1.MachineDeployment{
+		Spec: clusterv1.MachineDeploymentSpec{
+			Replicas: ptr.To[int32](0),
+			Template: clusterv1.MachineTemplateSpec{
+				Spec: clusterv1.MachineSpec{
+					Bootstrap: clusterv1.Bootstrap{
+						ConfigRef: &corev1.ObjectReference{
+							Kind:      "KubeadmBootstrapTemplate",
+							Namespace: "some-namespace",
+							Name:      "some-name",
+						},
+					},
+					InfrastructureRef: corev1.ObjectReference{
+						Kind:      "DockerMachineTemplate",
+						Namespace: "some-namespace",
+						Name:      "some-name",
+					},
+				},
+			},
+		},
+	}
+
+	scalingUpMachineDeploymentWith3Replicas := defaultMachineDeployment.DeepCopy()
+	scalingUpMachineDeploymentWith3Replicas.Spec.Replicas = ptr.To[int32](3)
+
+	deletingMachineDeploymentWith3Replicas := defaultMachineDeployment.DeepCopy()
+	deletingMachineDeploymentWith3Replicas.DeletionTimestamp = ptr.To(metav1.Now())
+	deletingMachineDeploymentWith3Replicas.Spec.Replicas = ptr.To[int32](3)
+
+	tests := []struct {
+		name                                         string
+		machineDeployment                            *clusterv1.MachineDeployment
+		machineSets                                  []*clusterv1.MachineSet
+		bootstrapTemplateNotFound                    bool
+		infrastructureTemplateNotFound               bool
+		getAndAdoptMachineSetsForDeploymentSucceeded bool
+		expectCondition                              metav1.Condition
+	}{
+		{
+			name:                           "getAndAdoptMachineSetsForDeploymentSucceeded failed",
+			machineDeployment:              defaultMachineDeployment,
+			bootstrapTemplateNotFound:      false,
+			infrastructureTemplateNotFound: false,
+			getAndAdoptMachineSetsForDeploymentSucceeded: false,
+			expectCondition: metav1.Condition{
+				Type:    clusterv1.MachineDeploymentScalingUpV1Beta2Condition,
+				Status:  metav1.ConditionUnknown,
+				Reason:  clusterv1.MachineDeploymentScalingUpInternalErrorV1Beta2Reason,
+				Message: "Please check controller logs for errors",
+			},
+		},
+		{
```

Comment:
Should we add a test case for MachineDeploymentScalingUpWaitingForReplicasSetV1Beta2Reason? Simila for scaling down